theseus/core/vectorizer.py

```diff
@@ -0,0 +1,301 @@
+        if self._cached_error is None:
+            return super().weighted_error()
+        return self._cached_error
+
+    def weighted_jacobians_error(
+        self,
+    ) -> Tuple[List[torch.Tensor], torch.Tensor]:
+        if self._cached_jacobians is None:
+            return super().weighted_jacobians_error()
+        assert self._cached_error is not None
+        return self._cached_jacobians, self._cached_error
+
+    def dim(self) -> int:
+        return self.cost_fn.dim()
+
+    def _copy_impl(self, new_name: Optional[str] = None) -> "_CostFunctionWrapper":
+        return _CostFunctionWrapper(self.cost_fn, name=new_name)
+
+    def __str__(self) -> str:
+        return f"Wrapper({str(_get_cost_function_schema(self.cost_fn))})"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+# This class replaces the Objective's iterator for one that takes advantage of
+# cost function vectorization
+# TODO:
+#   - Tests to add:
+#         + Test that vectorization results in correct costs
+#         + Vectorize variable update after NLOPT step
+class Vectorize:
+    _SHARED_TOKEN = "__shared__"
+
+    def __init__(self, objective: Objective):
+        # Each cost function is assigned a wrapper. The wrapper will hold the error
+        # and jacobian after vectorization.
+        self._cost_fn_wrappers: List[_CostFunctionWrapper] = []
+
+        # This dict groups cost functions that can be vectorized together.
+        self._schema_dict: Dict[
+            _CostFunctionSchema, List[_CostFunctionWrapper]
+        ] = defaultdict(list)
+
+        # Create wrappers for all cost functions and also get their schemas
+        for cost_fn in objective.cost_functions.values():
+            wrapper = _CostFunctionWrapper(cost_fn)
+            self._cost_fn_wrappers.append(wrapper)
+            schema = _get_cost_function_schema(cost_fn)
+            self._schema_dict[schema].append(wrapper)
+
+        # Now create a vectorized cost function for each unique schema
+        self._vectorized_cost_fns: Dict[_CostFunctionSchema, CostFunction] = {}
+        for schema in self._schema_dict:
+            base_cost_fn = self._schema_dict[schema][0].cost_fn
+            vectorized_cost_fn = base_cost_fn.copy(keep_variable_names=False)
+            vectorized_cost_fn.weight = base_cost_fn.weight.copy(
+                keep_variable_names=False
+            )
+            self._vectorized_cost_fns[schema] = vectorized_cost_fn
+
+        # Dict[_CostFunctionSchema, List[str]]
+        self._var_names = self._get_var_names()
+
+        # `vectorize()` will compute an error vector for each schema, then populate
+        # the wrappers with their appropriate weighted error slice.
+        # Replacing `obj._cost_functions_iterable` allows to recover these when
+        # iterating the Objective.
+        objective._cost_functions_iterable = self._cost_fn_wrappers
+        objective._vectorization_run = self._vectorize
```

Comment:
self._vectorize doesn't yet seem to be defined in the init function or anywhere else in this file.